### PR TITLE
fix(reader): adding in a reader url resolver

### PR DESCRIPTION
**Pull request by bassrock**
# Goal

When a `getpocket.com/read/[slug]` is requested from the parser, we should attempt to resolve it to the original item object. This will ensure that iOS, Android and Web get the right metadata in the saves view and all other item queries.

https://mozilla-hub.atlassian.net/browse/POCKET-10000

### CODE REVIEW

File: servers/parser-graphql-wrapper/src/apollo/resolvers.ts [hunk 1]
```diff
@@ -248,3 +229,34 @@ export const resolvers: Resolvers = {
...
 };
+
+/**
+ * Helper function to resolve an item from a url.
+ * This exists because we do a bit of logic depending on if the URL is a pocket owned url
+ * @param url
+ * @param context: IContext
+ * @returns an item or null
+ */
+const itemFromUrl = async (
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/apollo/resolvers.ts:240]: 👍

File: servers/parser-graphql-wrapper/src/readerView/readersSlug.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+/**
+ * Extract the e code from a URL. If the URL is a valid reader sharable link URL,
+ * return the encoded item id. Otherwise return undefined (e.g. if the URL is not a pocket reader URL).
+ * @param url the url to attempt to extract the encoded item id from
+ * @returns undefined if there is no slug to get or the encoded item id if it is a reader url
+ */
+export function extractSlugFromReadUrl(url: string): string | undefined {
+  const regex = /^(?:https?:\/\/)?getpocket\.com\/read\/((?!\d+$)[^/?#]+)/;
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/readerView/readersSlug.ts:29]: can you explain why you included special characters here? the ID should not contain them

Comment (kschelonka) [reply servers/parser-graphql-wrapper/src/readerView/readersSlug.ts:29]: I'm pretty sure this is what you want:

`/^(?:https?:\/\/)?getpocket\.com\/read\/([A-Za-z\d]+_[A-Za-z\d]+)$/gm`

Comment (kschelonka) [reply servers/parser-graphql-wrapper/src/readerView/readersSlug.ts:29]: With test cases: 
image

Comment (kschelonka) [reply servers/parser-graphql-wrapper/src/readerView/readersSlug.ts:29]: optionally allowing an ending `/` (not in capturing group)

`^(?:https?:\/\/)?getpocket\.com\/read\/([A-Za-z\d]+_[A-Za-z\d]+)\/?$`

Comment (kschelonka) [reply servers/parser-graphql-wrapper/src/readerView/readersSlug.ts]: resolving -- we did more regexing in slack

File: servers/parser-graphql-wrapper/src/readerView/readerSlug.spec.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      url: 'http://getpocket.com/read/12345123123123123789asdasdasdasdasdasd',
+      expected: undefined,
+    },
+    {
+      url: 'http://getpocket.com/read/fe562f9c5BCfC1eeQ9AffKeCaiD2a190J7E6a1f247B54Egd22_202cb962ac59075b964b07152d234b70?124=fa',
+      expected:
+        'fe562f9c5BCfC1eeQ9AffKeCaiD2a190J7E6a1f247B54Egd22_202cb962ac59075b964b07152d234b70',
+    },
+    {
+      url: 'http://getpocket.com/read/fe562f9c5BCfC1eeQ9AffKeCaiD2a190J7E6a1f247B54Egd22_202cb962ac59075b964b07152d234b70?utm_source=pocket-saves',
```
Comment (kschelonka) [inline servers/parser-graphql-wrapper/src/readerView/readerSlug.spec.ts:46]: add a couple more test cases:

repeat one of the above with no http/s (getpocket.com/read/...)

add one with multiple query params e.g. `getpocket.com/read/asdf_2343?utm_source=pocket-saves&page=123`

add one with a trailing `/` e.g. `getpocket.com/read/asdf_343/`

Comment (bassrock) [reply servers/parser-graphql-wrapper/src/readerView/readerSlug.spec.ts:46]: added!

**Review by kschelonka — APPROVED** [review]

**Pull request merged**